File: commons/services-util/src/main/java/com/enablix/services/util/template/walker/TemplateContainerWalker.java
package com.enablix.services.util.template.walker;

import java.util.List;

import com.enablix.core.commons.xsdtopojo.ContainerType;
import com.enablix.core.commons.xsdtopojo.ContentTemplate;

public class TemplateContainerWalker {

	private ContentTemplate template;
	private ContainerFilter filter;
	
	public TemplateContainerWalker(ContentTemplate template) {
		this(template, new VisitAllContainerFilter());
	}
	
	public TemplateContainerWalker(ContentTemplate template, ContainerFilter filter) {
		this.template = template;
		this.filter = filter;
	}
	
	public void walk(ContainerVisitor visitor) {
		walkAndProcess(template.getDataDefinition().getContainer(), visitor);
	}
	
	protected void walkAndProcess(List<ContainerType> containers, ContainerVisitor visitor) {
		
		for (ContainerType container : containers) {
			
			if (filter.accept(container)) {
				visitor.visit(container);
			}
			
			walkAndProcess(container.getContainer(), visitor);
		}
	}
	
}

File: web2/ebx-web/src/app/util/template-container-walker.ts
import { Container } from '../model/container.model';
 
export class TemplateContainerWalker {

  containers: Container[];

  constructor(containers: Container[]) {
    this.containers = containers;
  }

  walk(visitor: (container: Container) => void) {
    this.walkAndProcess(this.containers, visitor);
  }
  
  private walkAndProcess(containers: Container[], visitor: (container: Container) => void) {
    
    containers.forEach(container => {
      
      // visit this container
      visitor(container);

      // visit child containers
      this.walkAndProcess(container.container, visitor);
    });
    
  }
  
}